woocommerce/woocommerce-ios
# [SKU Scanner]Filter non-purchasable items out

Closes: #9989 


## Description

With this PR we filter out non-purchasable items when scanning them with the new barcode scanner feature. As they cannot be added to an order, we show a message to the user stating that error.

To fix this I opted to refactor the ProductStore function to filter these products. The rationale behind implementing this here in the store instead of later in the process is straightforward: we want to fail as soon as possible to the code simple and take advantage of the fact that the purchasable information is fully available before wrapping our item into a `SKUSearchResult` instance. We extended here the `ProductLoadError` to include this case, which might not be 100% ideal but that makes it very easy and convenient for us.

After that, I created a separate struct to generate the error notices (`BarcodeSKUScannerErrorNoticeFactory`) depending on the error, that is used now in both the Order List and Order Creation. To make it even simpler I unified the notice strings in both places.

## Testing instructions


Add a non purchasable (no price or draft) product or variation to your store. Assign it a SKU and generate a barcode for that one. You can use this [tool](https://barcode.tec-it.com/en/UPCA?data=72527273070).

1. Go to orders
2. Scan that product. 
3. See that a notice with title _Product not purchasable. Failed to add Product to Order._ is shown and the product is not added to the order.
4. Re-try with a valid product. See that it's added as usual



## Screenshots


https://github.com/woocommerce/woocommerce-ios/assets/1864060/14f4e125-1607-43a2-9680-e52d62e502ef

---
